File: app/src/main/java/io/github/micopiira/mylauncher/AppRepository.java
package io.github.micopiira.mylauncher;

import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import butterknife.BindString;

public class AppRepository implements CrudRepository<AppDetail> {

    private final PackageManager packageManager;
    private final Context context;
    private final ApplicationInfo applicationInfo;
    private final String appName;

    public AppRepository(Context context, String appName) {
        this.packageManager = context.getPackageManager();
        this.context = context;
        this.applicationInfo = context.getApplicationInfo();
        this.appName = appName;
    }

    @Override
    public long count() {
        return findAll().size();
    }

    @Override
    public List<AppDetail> findAll() {
        Intent i = new Intent(Intent.ACTION_MAIN, null);
        i.addCategory(Intent.CATEGORY_LAUNCHER);

        List<ResolveInfo> resolveInfoList = packageManager.queryIntentActivities(i, 0);
        List<AppDetail> apps = new ArrayList<>();
        for (ResolveInfo resolveInfo : resolveInfoList) {
            AppDetail app = new AppDetail();
            app.setLabel(resolveInfo.loadLabel(packageManager).toString());
            app.setName(resolveInfo.activityInfo.packageName);
            app.setIcon(resolveInfo.activityInfo.loadIcon(packageManager));
            app.setIntent(packageManager.getLaunchIntentForPackage(resolveInfo.activityInfo.packageName));
            apps.add(app);
        }
        AppDetail launcherSettings = new AppDetail();
        launcherSettings.setLabel(appName + " settings");
        launcherSettings.setIcon(packageManager.getApplicationIcon(applicationInfo));
        launcherSettings.setIntent(new Intent(context, PreferencesActivity.class));
        apps.add(launcherSettings);
        Collections.sort(apps, (o1, o2) -> o1.getLabel().compareToIgnoreCase(o2.getLabel()));
        return apps;
    }
}

File: app/src/main/java/io/github/micopiira/mylauncher/AppRepository.kt
package io.github.micopiira.mylauncher

import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager

class AppRepository(context: Context, appName: String) {

    private val packageManager: PackageManager = context.packageManager
    private val applicationInfo: ApplicationInfo = context.applicationInfo
    private val settingsApp = AppDetail(
            label = "$appName settings",
            icon = packageManager.getApplicationIcon(applicationInfo),
            intent = Intent(context, PreferencesActivity::class.java)
    )

    fun findAll(): List<AppDetail> {
        val i = Intent(Intent.ACTION_MAIN, null).addCategory(Intent.CATEGORY_LAUNCHER)

        return packageManager.queryIntentActivities(i, 0).map {
            AppDetail(
                    label = it.loadLabel(packageManager).toString(),
                    name = it.activityInfo.packageName,
                    icon = it.activityInfo.loadIcon(packageManager),
                    intent = packageManager.getLaunchIntentForPackage(it.activityInfo.packageName)
            )
        }.plus(settingsApp).sortedBy { it.label }
    }
}
